core/src/main/java/org/elasticsearch/index/translog/TruncateTranslogCommand.java

```diff
@@ -0,0 +1,213 @@
+        } catch (IOException e) {
+            terminal.println("encountered IOException while listing directory, aborting...");
+            throw new ElasticsearchException("failed to find existing translog files", e);
+        }
+
+        // Retrieve the generation and UUID from the existing data
+        Path idxLocation = translogPath.getParent().resolve("index");
+        terminal.println("Reading translog UUID information from Lucene commit from shard at [" + idxLocation + "]");
+        Map<String, String> commitData = readCommitInfo(idxLocation);
+        String translogGeneration = commitData.get(Translog.TRANSLOG_GENERATION_KEY);
+        String translogUUID = commitData.get(Translog.TRANSLOG_UUID_KEY);
+        if (translogGeneration == null || translogUUID == null) {
+            throw new ElasticsearchException("shard must have a valid translog generation and UUID but got: [{}] and: [{}]",
+                    translogGeneration, translogUUID);
+        }
+        terminal.println("Translog Generation: " + translogGeneration);
+        terminal.println("Translog UUID      : " + translogUUID);
+
+        Path tempEmptyCheckpoint = translogPath.resolve("temp-" + Translog.CHECKPOINT_FILE_NAME);
+        Path realEmptyCheckpoint = translogPath.resolve(Translog.CHECKPOINT_FILE_NAME);
+        Path tempEmptyTranslog = translogPath.resolve("temp-" + Translog.TRANSLOG_FILE_PREFIX +
+                        translogGeneration + Translog.TRANSLOG_FILE_SUFFIX);
+        Path realEmptyTranslog = translogPath.resolve(Translog.TRANSLOG_FILE_PREFIX + translogGeneration + Translog.TRANSLOG_FILE_SUFFIX);
+
+        // Write empty checkpoint and translog to empty files
+        long gen = Long.parseLong(translogGeneration);
+        writeEmptyCheckpoint(tempEmptyCheckpoint, gen);
+        writeEmptyTranslog(tempEmptyTranslog, translogUUID);
+
+        terminal.println("Removing existing translog files");
+        IOUtils.rm(translogFiles.toArray(new Path[]{}));
+
+        terminal.println("Creating new empty checkpoint at [" + realEmptyCheckpoint + "]");
+        Files.move(tempEmptyCheckpoint, realEmptyCheckpoint, StandardCopyOption.ATOMIC_MOVE);
+        terminal.println("Creating new empty translog at [" + realEmptyTranslog + "]");
+        Files.move(tempEmptyTranslog, realEmptyTranslog, StandardCopyOption.ATOMIC_MOVE);
+
+        terminal.println("Done.");
+    }
+
+    /**
+     * Read the commit info map from the given shard location by opening an {@code IndexWriter}, retrieving the data, then immediately
+     * closing it.
+     */
+    public static Map<String, String> readCommitInfo(Path location) throws Exception {
+        final IndexWriterConfig iwc = new IndexWriterConfig(null);
+        iwc.setOpenMode(IndexWriterConfig.OpenMode.APPEND);
+        try (Directory dir = FSDirectory.open(location, NativeFSLockFactory.INSTANCE)) {
+            try (IndexWriter iw = new IndexWriter(dir, iwc)) {
```

Comment:
Instead of using `IndexWriter` here, you could use `DirectoryReader.listCommits` (there should be at most `IndexCommit` returned) and then call `IndexCommit.getUserData()` instead.  Seems safer since `DirectoryReader` cannot do any writing on the index, doesn't acquire the write lock, etc.